### PR TITLE
UsbSerialDiscovery service based on Windows registry

### PR DESCRIPTION
**Pull request by andrewfg**
This is a new implementation of the UsbSerialDiscovery interface for Windows, which discovers USB devices by reading from the Windows registry.

This may be an alternative, or an extension, to #3930 

Signed-off-by: Andrew Fiddian-Green [software@whitebear.ch](mailto:software@whitebear.ch)

### CODE REVIEW

**Comment by andrewfg** [comment]
@mherwege 4yi

**Comment by andrewfg** [comment]
@mherwege it will probably fail the Karaf build (as usual) due to missing feature :(

**Comment by andrewfg** [comment]
@mherwege I added the missing serial port and interface information

**Comment by andrewfg** [comment]
As https://github.com/openhab/openhab-core/pull/3922 is now merged, this one is also ready for review.

**Review by wborn — CHANGES_REQUESTED** [review]
This seems useful for Windows users. :+1: 

I only have a few minor comments:

File: bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/test/java/org/openhab/core/config/discovery/usbserial/windowsregistry/test/WindowsUsbDiscoveryTests.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * JUnit tests for {@link WindowsUsbSerialDiscovery}
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+class WindowsUsbDiscoveryTests {
+
+    @Test
+    void testDiscovery() {
+        // TODO write tests
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/test/java/org/openhab/core/config/discovery/usbserial/windowsregistry/test/WindowsUsbDiscoveryTests.java:29]: Umm. Not yet. I am still trying to figure out how to mock it in a meaningful way. I think the only real test is a live test on an actual (Windows) system with an actual stick plugged in. And I did that test myself already. So a mockito test would be nice-to-have but perhaps not essential. => @wborn WDYT?

File: bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/main/java/org/openhab/core/config/discovery/usbserial/windowsregistry/internal/WindowsUsbSerialDiscovery.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+        Set<UsbSerialDeviceInformation> scanResult = scanAllUsbDevicesInformation();
+        Set<UsbSerialDeviceInformation> added = setDifference(scanResult, lastScanResult);
+        Set<UsbSerialDeviceInformation> removed = setDifference(lastScanResult, scanResult);
+        Set<UsbSerialDeviceInformation> unchanged = setDifference(scanResult, added);
+
+        lastScanResult = scanResult;
+
+        removed.stream().forEach(this::announceRemovedDevice);
+        added.stream().forEach(this::announceAddedDevice);
+        unchanged.stream().forEach(this::announceAddedDevice);
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/main/java/org/openhab/core/config/discovery/usbserial/windowsregistry/internal/WindowsUsbSerialDiscovery.java:111]: @wborn I will make the changes in this class. But I think there should be analog changes in the class from which I copied the code => do you want me to make those changes in this PR or another?

File: bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/main/java/org/openhab/core/config/discovery/usbserial/windowsregistry/internal/WindowsUsbSerialDiscovery.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        }
+        return result;
+    }
+
+    @Override
+    public synchronized void startBackgroundScanning() {
+        if (Platform.isWindows()) {
+            ScheduledFuture<?> scanTask = this.scanTask;
+            if (scanTask == null || scanTask.isDone()) {
+                this.scanTask = scheduler.scheduleWithFixedDelay(() -> doSingleScan(), 0, scanInterval.toSeconds(),
```
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/main/java/org/openhab/core/config/discovery/usbserial/windowsregistry/internal/WindowsUsbSerialDiscovery.java:247]: Ok.

File: bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/main/java/org/openhab/core/config/discovery/usbserial/windowsregistry/internal/WindowsUsbSerialDiscovery.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/main/java/org/openhab/core/config/discovery/usbserial/windowsregistry/internal/WindowsUsbSerialDiscovery.java]: https://github.com/openhab/openhab-core/blob/2abe4e20ed1133c89afe43179694c8204d3e6e35/bundles/org.openhab.core.config.discovery.usbserial.ser2net/src/main/java/org/openhab/core/config/discovery/usbserial/ser2net/internal/Ser2NetUsbSerialDiscovery.java#L128

File: bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/test/java/org/openhab/core/config/discovery/usbserial/windowsregistry/test/WindowsUsbDiscoveryTests.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (wborn) [reply bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/test/java/org/openhab/core/config/discovery/usbserial/windowsregistry/test/WindowsUsbDiscoveryTests.java:29]: You could move all the `com.sun.jna` logic into some helper class. That class could be mocked and then you can easily test the logic in a unit test. If you don't do this you should remove the useless test class.

File: bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/main/java/org/openhab/core/config/discovery/usbserial/windowsregistry/internal/WindowsUsbSerialDiscovery.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (wborn) [reply bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/main/java/org/openhab/core/config/discovery/usbserial/windowsregistry/internal/WindowsUsbSerialDiscovery.java]: If there aren't many changes you can do it in this PR.

File: bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/test/java/org/openhab/core/config/discovery/usbserial/windowsregistry/test/WindowsUsbDiscoveryTests.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewfg) [reply bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/src/test/java/org/openhab/core/config/discovery/usbserial/windowsregistry/test/WindowsUsbDiscoveryTests.java:29]: I will remove the test case ;)

**Comment by andrewfg** [comment]
@wborn there is still the question of which UsbSerialDiscovery components should be loaded on which platforms .. I guess the following .. => WDYT? And the next question is how to program the dependencies and karaf stuff .. which is probably beyond my capabilities :(

- ser2net - all platforms
- sysfs - Linux
- windowsregistry - Windows
- javaxusb - all platforms except Linux and Windows

**Comment by wborn** [comment]
>  I guess the following

Yes that probably makes sense.

> sysfs - Linux

It doesn't use any native dependency which is very convenient because Linux runs on so many platforms. So it will result in fewer issues compared to javaxusb.

> javaxusb - all platforms except Linux and Windows

It uses native libraries and only supports Linux, Windows and macOS. So it probably only makes sense to use it on macOS.

> And the next question is how to program the dependencies and karaf stuff

You may be able to use `<conditional>` for that, see also https://github.com/openhab/openhab-core/pull/3632#issuecomment-1592847775.

**Comment by andrewfg** [comment]
> may be able to use <conditional> for that

@wborn the following is simply a wild guess idea .. but I would like to have some feedback from more knowledgeable Karaf experts before even attempting to make such a commit.

.. in particular if the `osgi.native.osname~=Windows` is a proper syntax..

```
<feature name="openhab-transport-serial" description="Serial Transport" version="${project.version}">
    <feature>openhab-core-base</feature>

    <requirement>openhab.tp;filter:="(feature=commons-net)"</requirement>
    <feature dependency="true">openhab.tp-commons-net</feature>

    <requirement>openhab.tp;filter:="(&amp;(feature=serial)(impl=rxtx))"</requirement>
    <feature dependency="true">openhab.tp-serial-rxtx</feature>

    <feature dependency="true">openhab-core-io-transport-mdns</feature>

    <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.serial/${project.version}</bundle>
    <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.discovery.usbserial/${project.version}</bundle>
    
    <conditional>
        <condition>req:osgi.native;filter:="(osgi.native.osname=MacOS)"</condition>
        <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.discovery.usbserial.javaxusb/${project.version}</bundle>
    </conditional>
    
    <conditional>
        <condition>req:osgi.native;filter:="(osgi.native.osname=Linux)"</condition>
        <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.discovery.usbserial.linuxsysfs/${project.version}</bundle>
    </conditional>
    
    <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.discovery.usbserial.ser2net/${project.version}</bundle>
    
    <conditional>
        <condition>req:osgi.native;filter:="(osgi.native.osname~=Windows)"</condition>
        <bundle>mvn:org.openhab.core.bundles/org.openhab.core.config.discovery.usbserial.windowsregistry/${project.version}</bundle>
    </conditional>
    
    <bundle>mvn:org.openhab.core.bundles/org.openhab.core.io.transport.serial/${project.version}</bundle>
    <bundle>mvn:org.openhab.core.bundles/org.openhab.core.io.transport.serial.rxtx/${project.version}</bundle>
    <bundle>mvn:org.openhab.core.bundles/org.openhab.core.io.transport.serial.rxtx.rfc2217/${project.version}</bundle>

</feature>
```

**Comment by wborn** [comment]
I haven't this myself yet. Just give it a try and see what happens. :slightly_smiling_face:

**Comment by andrewfg** [comment]
> Just give it a try and see what happens

Yeah. I just ran a full build of OH this afternoon (it takes ages on my machine; I probably need to buy a new one). The build succeeded without error. But I will test it tomorrow.

**Comment by andrewfg** [comment]
@wborn below is a screenshot in Windows showing that this (Windows) discovery is installed, and the existing (Linux) discovery is not. To be ultra safe we would need to check the opposite case on a Linux machine. However I currently don't have a Linux test machine, so I wonder if you could eventually test it for me?

image

**Comment by wborn** [comment]
> I wonder if you could eventually test it for me?

I just gave it a test on Ubuntu and the linuxsysfs bundle keeps getting installed while the windowsregistry bundle is not. :+1:

**Review by wborn — APPROVED** [review]
Thanks for the new UsbSerialDiscovery!

**Comment by andrewfg** [comment]
> gave it a test on Ubuntu and the linuxsysfs bundle keeps getting installed

I also -- finally -- managed to test it on a RasPi3 running Linux and it also shows the correct components loaded..

image

**Pull request merged**